File: leetcode_practice/typescript/leetcode_101_test.ts
/*
 * LeetCode #101: 题目101
 * 难度: 未知
 * 
 * 题目描述:
 * 由大模型直接生成
 * 
 * 代码骨架完整度: 10%
 */

// LeetCode #101 对称二叉树 (Symmetric Tree)
// 难度：简单
// 题目描述：给定一个二叉树的根节点root，检查它是否轴对称（即镜像对称）。

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

// 解法1：递归法
// 算法思路：比较左右子树是否镜像对称，即左子树的左节点等于右子树的右节点，左子树的右节点等于右子树的左节点
// 时间复杂度：O(n)，需要遍历所有节点
function isSymmetric(root: TreeNode | null): boolean {
    // TODO: 实现递归逻辑
    // 提示：可以定义一个辅助函数来递归比较两个节点
    return false;
}

// 解法2：迭代法
// 算法思路：使用队列或栈来模拟递归过程，每次取出两个节点比较
// 时间复杂度：O(n)，需要遍历所有节点
function isSymmetricIterative(root: TreeNode | null): boolean {
    // TODO: 实现迭代逻辑
    // 提示：可以使用队列，每次入队两个对称节点进行比较
    return false;
}

// 测试用例
function TestIsSymmetric() {
    // 测试用例1：对称二叉树
    //     1
    //    / \
    //   2   2
    //  / \ / \
    // 3  4 4  3
    const tree1 = new TreeNode(1,
        new TreeNode(2, new TreeNode(3), new TreeNode(4)),
        new TreeNode(2, new TreeNode(4), new TreeNode(3))
    );
    console.assert(isSymmetric(tree1) === true, "测试用例1失败");

    // 测试用例2：不对称二叉树
    //     1
    //    / \
    //   2   2
    //    \   \
    //    3    3
    const tree2 = new TreeNode(1,
        new TreeNode(2, null, new TreeNode(3)),
        new TreeNode(2, null, new TreeNode(3))
    );
    console.assert(isSymmetric(tree2) === false, "测试用例2失败");
}

// 比较函数（用于测试）
function compareTrees(p: TreeNode | null, q: TreeNode | null): boolean {
    if (p === null && q === null) return true;
    if (p === null || q === null) return false;
    return p.val === q.val && compareTrees(p.left, q.left) && compareTrees(p.right, q.right);
}

// 运行测试
TestIsSymmetric();

File: leetcode_practice/go/leetcode_101_test.go
/*
 * LeetCode #101: 题目101
 * 难度: 未知
 * 
 * 题目描述:
 * 由大模型直接生成
 * 
 * 代码骨架完整度: 10%
 */

package leetcode

import (
	"testing"
)

/*
题目：#101 对称二叉树（Symmetric Tree）
难度：简单
描述：给定一个二叉树的根节点root，检查它是否轴对称。
*/

// TreeNode 定义二叉树节点
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

/*
解法一：递归法
算法思路：
1. 对称二叉树的条件是左子树和右子树互为镜像
2. 两个树互为镜像的条件：
   - 它们的根节点值相同
   - 每个树的左子树与另一个树的右子树互为镜像
时间复杂度：O(n)，需要遍历所有节点
空间复杂度：O(h)，h为树的高度，递归调用栈的深度
*/
func isSymmetric(root *TreeNode) bool {
	// TODO: 实现递归判断逻辑
	// 提示：可以定义一个辅助函数来比较两个子树是否对称
	if root == nil {
		return true
	}
	var isMirror func(t1, t2 *TreeNode) bool
	isMirror = func(t1, t2 *TreeNode) bool {
		if t1 == nil && t2 == nil {
			return true
		}
		if t1 == nil || t2 == nil {
			return false
		}
		return t1.Val == t2.Val && isMirror(t1.Left, t2.Right) && isMirror(t1.Right, t2.Left)
	}
	return isMirror(root.Left, root.Right)
}

/*
解法二：迭代法（使用队列）
算法思路：
1. 使用队列来按层比较节点
2. 每次从队列中取出两个节点比较它们的值
3. 将左子树的左节点和右子树的右节点入队，左子树的右节点和右子树的左节点入队
时间复杂度：O(n)，需要遍历所有节点
空间复杂度：O(n)，最坏情况下队列需要存储所有叶子节点
*/
func isSymmetricIterative(root *TreeNode) bool {
	// TODO: 实现迭代判断逻辑
	// 提示：使用队列或栈来存储需要比较的节点对
	return false
}

// 测试用例
func TestIsSymmetric(t *testing.T) {
	tests := []struct {
		name string
		root *TreeNode
		want bool
	}{
		{
			name: "对称二叉树",
			root: &TreeNode{
				Val: 1,
				Left: &TreeNode{
					Val:   2,
					Left:  &TreeNode{Val: 3},
					Right: &TreeNode{Val: 4},
				},
				Right: &TreeNode{
					Val:   2,
					Left:  &TreeNode{Val: 4},
					Right: &TreeNode{Val: 3},
				},
			},
			want: true,
		},
		{
			name: "非对称二叉树",
			root: &TreeNode{
				Val: 1,
				Left: &TreeNode{
					Val:   2,
					Right: &TreeNode{Val: 3},
				},
				Right: &TreeNode{
					Val:   2,
					Right: &TreeNode{Val: 3},
				},
			},
			want: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := isSymmetric(tt.root); got != tt.want {
				t.Errorf("isSymmetric() = %v, want %v", got, tt.want)
			}
			// if got := isSymmetricIterative(tt.root); got != tt.want {
			// 	t.Errorf("isSymmetricIterative() = %v, want %v", got, tt.want)
			// }
		})
	}
}
